include/revng/ADT/TrackingContainer.h

```diff
@@ -0,0 +1,241 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <stack>
+#include <vector>
+
+#include "revng/Support/AccessCounter.h"
+
+namespace revng {
+template<typename T>
+class TrackingContainer {
+private:
+  T Content;
+
+public:
+  using key_type = typename T::key_type;
+
+private:
+  using TrackingVector = std::vector<std::remove_const_t<key_type>>;
+  mutable AccessCounter Exact;
+  mutable std::stack<TrackingVector> NonExisting;
+  mutable std::stack<TrackingVector> Existing;
```

Comment:
i guess i will go for std::set and if we later need to i will upgrade to something more efficient

Reply:
Please give a shot to the bit storage too, I see no reason to try to be super efficient with scalar fields using bits magic but not doing it for `TrackingContainer`.

Is there an actual reason for your reluctance or is it actually just because of the lack of symmetry?

Reply:
For the inspector methods: I think we can go even more aggressive with hiding implementation details and have a single inspector that flattens everything out and returns a `std::set<key>` merging the `Exact`, `Existing` and `NonExisting`.

If we do like this, the user won't notice the asymmetry and we can do the bit thing!

Reply:
yeah, we can go that way

Reply:
I still see the three methods...